packages/legend-query/src/stores/QueryBuilderMilestoningHelper.ts

```diff
@@ -0,0 +1,207 @@
+      param1.genericType.value.rawType.name ===
+      param2.genericType.value.rawType.name
+    ) {
+      return param1.values[0] === param2.values[0];
+    }
+    return false;
+  }
+  return false;
+};
+
+export const getSourceTemporalStereotype = (
+  property: DerivedProperty,
+  graph: PureModel,
+): MILESTONING_STEREOTYPE | undefined => {
+  const owner = property.owner;
+  if (owner instanceof Class) {
+    return getMilestoneTemporalStereotype(owner, graph);
+  } else if (owner instanceof Association) {
+    if (owner._generatedMilestonedProperties.length) {
+      const ownerClass =
+        owner._generatedMilestonedProperties[0]?.genericType.value.rawType;
+      return getMilestoneTemporalStereotype(
+        guaranteeType(ownerClass, Class),
+        graph,
+      );
+    }
+  }
+  return undefined;
+};
+
+export const checkForOverridingMilestoningQuery = (
+  derivedPropertyExpressionState: QueryBuilderDerivedPropertyExpressionState,
+  targetStereotype: MILESTONING_STEREOTYPE,
+): boolean => {
+  const sourceStereotype = getSourceTemporalStereotype(
+    derivedPropertyExpressionState.derivedProperty,
+    derivedPropertyExpressionState.queryBuilderState.graphManagerState.graph,
+  );
+  if (
+    sourceStereotype === MILESTONING_STEREOTYPE.PROCESSING_TEMPORAL &&
+    targetStereotype === MILESTONING_STEREOTYPE.BITEMPORAL &&
+    derivedPropertyExpressionState.parameterValues.length === 1
+  ) {
+    return true;
+  } else {
+    return false;
+  }
+};
+
+/**
+ * Checks if the parameter of the current milestoned property is same as `businessDate` or `processingDate`
+ */
+export const checkForMilestoningParameterEquality = (
+  stereotype: MILESTONING_STEREOTYPE,
+  idx: number,
+  milestoningParameter: ValueSpecification,
+  querySetupState: QueryBuilderSetupState,
+): boolean => {
+  switch (stereotype) {
+    case MILESTONING_STEREOTYPE.BITEMPORAL:
+      if (
+        idx === 0 &&
+        valueSpecifiation_isEqual(
+          milestoningParameter,
+          querySetupState.processingDate,
+        )
+      ) {
+        return true;
+      } else if (
+        idx === 1 &&
+        valueSpecifiation_isEqual(
+          milestoningParameter,
+          querySetupState.businessDate,
+        )
+      ) {
+        return true;
+      } else {
+        return false;
+      }
+    case MILESTONING_STEREOTYPE.PROCESSING_TEMPORAL:
+      if (
+        valueSpecifiation_isEqual(
+          milestoningParameter,
+          querySetupState.processingDate,
+        )
+      ) {
+        return true;
+      } else {
+        return false;
+      }
+    case MILESTONING_STEREOTYPE.BUSINESS_TEMPORAL:
+      if (
+        valueSpecifiation_isEqual(
+          milestoningParameter,
+          querySetupState.businessDate,
+        )
+      ) {
+        return true;
+      } else {
+        return false;
+      }
+    default:
+  }
+  return false;
+};
+
+/**
+ * sourceStereotype: Stereotype of source class of current property expression.
```

Comment:
the name of the method speaks for itself really, but I think we should move these comment to where we create these variables instead. As this is useful for code readability, not as documentation of the function's usage.